File: server_csharp/Network/Protocol.cs
using System;
using System.IO;
using System.Text;

namespace AnantaServer.Network
{
    public static class PacketModes
    {
        public const byte Invoke = 1;
        public const byte Return = 2;
        public const byte Notify = 3;

        public const byte MarkNull = 0;
        public const byte MarkCommon = 255;
    }

    public class RpcPacket
    {
        public byte Mode { get; set; }
        public int MethodId { get; set; }
        public int InvokeId { get; set; }
        public int ErrId { get; set; }
        public byte[] Payload { get; set; } = Array.Empty<byte>();

        public RpcPacket(byte mode, int methodId, int invokeId = 0, int errId = 0, byte[]? payload = null)
        {
            Mode = mode;
            MethodId = methodId;
            InvokeId = invokeId;
            ErrId = errId;
            Payload = payload ?? Array.Empty<byte>();
        }

        public byte[] Encode(bool includeFraming = true)
        {
            using var ms = new MemoryStream();
            using var bw = new BinaryWriter(ms);

            bw.Write(Mode);
            bw.Write(MethodId);

            if (Mode == PacketModes.Invoke || Mode == PacketModes.Return)
            {
                bw.Write(InvokeId);
            }

            if (Mode == PacketModes.Return)
            {
                bw.Write(ErrId);
            }

            if (Payload.Length > 0)
            {
                bw.Write(Payload);
            }

            var body = ms.ToArray();
            if (!includeFraming)
            {
                return body;
            }

            using var framed = new MemoryStream();
            using var fbw = new BinaryWriter(framed);
            fbw.Write((uint)body.Length);
            fbw.Write(body);
            return framed.ToArray();
        }

        public static RpcPacket Decode(byte[] data)
        {
            if (data.Length < 5)
                throw new ArgumentException($"Packet too short: {data.Length} bytes");

            using var ms = new MemoryStream(data);
            using var br = new BinaryReader(ms);

            byte mode = br.ReadByte();
            int methodId = br.ReadInt32();
            int invokeId = 0;
            int errId = 0;

            if (mode == PacketModes.Invoke || mode == PacketModes.Return)
            {
                invokeId = br.ReadInt32();
            }

            if (mode == PacketModes.Return)
            {
                errId = br.ReadInt32();
            }

            int remaining = (int)(ms.Length - ms.Position);
            byte[] payload = remaining > 0 ? br.ReadBytes(remaining) : Array.Empty<byte>();

            return new RpcPacket(mode, methodId, invokeId, errId, payload);
        }
    }

    public static class UxBinaryExtensions
    {
        public static void WriteUxString(this BinaryWriter writer, string? val, bool nullable = false)
        {
            if (val == null)
            {
                if (nullable)
                    writer.Write(PacketModes.MarkNull);
                else
                    writer.Write7BitEncodedInt(0);
                return;
            }

            if (nullable)
                writer.Write(PacketModes.MarkCommon);

            byte[] bytes = Encoding.UTF8.GetBytes(val);
            writer.Write7BitEncodedInt(bytes.Length);
            writer.Write(bytes);
        }

        public static string ReadUxString(this BinaryReader reader, bool nullable = false)
        {
            if (nullable)
            {
                byte mark = reader.ReadByte();
                if (mark == PacketModes.MarkNull)
                    return string.Empty;
            }

            int len = reader.Read7BitEncodedInt();
            if (len <= 0) return string.Empty;

            byte[] bytes = reader.ReadBytes(len);
            return Encoding.UTF8.GetString(bytes);
        }
    }
}

File: server_nodejs/network/protocol.js
/**
 * NetEase UX RPC Protocol Codec for Node.js
 */

const MODE_INVOKE = 1;
const MODE_RETURN = 2;
const MODE_NOTIFY = 3;

const MARK_NULL = 0;
const MARK_COMMON = 255;

class BinaryWriter {
    constructor() {
        this.buffers = [];
    }

    writeByte(val) {
        const b = Buffer.alloc(1);
        b.writeUInt8(val & 0xFF, 0);
        this.buffers.push(b);
    }

    writeBoolean(val) {
        this.writeByte(val ? 1 : 0);
    }

    writeInt16(val) {
        const b = Buffer.alloc(2);
        b.writeInt16LE(val, 0);
        this.buffers.push(b);
    }

    writeUInt16(val) {
        const b = Buffer.alloc(2);
        b.writeUInt16LE(val, 0);
        this.buffers.push(b);
    }

    writeInt32(val) {
        const b = Buffer.alloc(4);
        b.writeInt32LE(val, 0);
        this.buffers.push(b);
    }

    writeUInt32(val) {
        const b = Buffer.alloc(4);
        b.writeUInt32LE(val, 0);
        this.buffers.push(b);
    }

    writeInt64(val) {
        const b = Buffer.alloc(8);
        b.writeBigInt64LE(BigInt(val), 0);
        this.buffers.push(b);
    }

    writeUInt64(val) {
        const b = Buffer.alloc(8);
        b.writeBigUInt64LE(BigInt(val), 0);
        this.buffers.push(b);
    }

    writeSingle(val) {
        const b = Buffer.alloc(4);
        b.writeFloatLE(val, 0);
        this.buffers.push(b);
    }

    writeDouble(val) {
        const b = Buffer.alloc(8);
        b.writeDoubleLE(val, 0);
        this.buffers.push(b);
    }

    write7BitEncodedInt(val) {
        let v = Number(val) >>> 0;
        const bytes = [];
        while (v >= 0x80) {
            bytes.push((v | 0x80) & 0xFF);
            v >>>= 7;
        }
        bytes.push(v & 0xFF);
        this.buffers.push(Buffer.from(bytes));
    }

    writeString(val, nullable = false) {
        if (val === null || val === undefined) {
            if (nullable) {
                this.writeByte(MARK_NULL);
            } else {
                this.write7BitEncodedInt(0);
            }
            return;
        }

        if (nullable) {
            this.writeByte(MARK_COMMON);
        }

        const strBuf = Buffer.from(String(val), "utf-8");
        this.write7BitEncodedInt(strBuf.length);
        this.buffers.push(strBuf);
    }

    getBytes() {
        return Buffer.concat(this.buffers);
    }
}

class BinaryReader {
    constructor(buffer) {
        this.buffer = buffer;
        this.offset = 0;
    }

    remaining() {
        return this.buffer.length - this.offset;
    }

    readByte() {
        const val = this.buffer.readUInt8(this.offset);
        this.offset += 1;
        return val;
    }

    readBoolean() {
        return this.readByte() !== 0;
    }

    readInt16() {
        const val = this.buffer.readInt16LE(this.offset);
        this.offset += 2;
        return val;
    }

    readUInt16() {
        const val = this.buffer.readUInt16LE(this.offset);
        this.offset += 2;
        return val;
    }

    readInt32() {
        const val = this.buffer.readInt32LE(this.offset);
        this.offset += 4;
        return val;
    }

    readUInt32() {
        const val = this.buffer.readUInt32LE(this.offset);
        this.offset += 4;
        return val;
    }

    readInt64() {
        const val = this.buffer.readBigInt64LE(this.offset);
        this.offset += 8;
        return val;
    }

    readUInt64() {
        const val = this.buffer.readBigUInt64LE(this.offset);
        this.offset += 8;
        return val;
    }

    readSingle() {
        const val = this.buffer.readFloatLE(this.offset);
        this.offset += 4;
        return val;
    }

    readDouble() {
        const val = this.buffer.readDoubleLE(this.offset);
        this.offset += 8;
        return val;
    }

    read7BitEncodedInt() {
        let count = 0;
        let shift = 0;
        while (shift < 35) {
            const b = this.readByte();
            count |= (b & 0x7F) << shift;
            shift += 7;
            if ((b & 0x80) === 0) {
                return count;
            }
        }
        throw new Error("Invalid 7-bit encoded integer");
    }

    readString(nullable = false) {
        if (nullable) {
            const mark = this.readByte();
            if (mark === MARK_NULL) return null;
        }

        const len = this.read7BitEncodedInt();
        if (len === 0) return "";

        const str = this.buffer.toString("utf-8", this.offset, this.offset + len);
        this.offset += len;
        return str;
    }
}

class RpcPacket {
    constructor(mode, methodId, invokeId = 0, errId = 0, payload = Buffer.alloc(0)) {
        this.mode = mode;
        this.methodId = methodId;
        this.invokeId = invokeId;
        this.errId = errId;
        this.payload = payload;
    }

    encode(includeFraming = true) {
        const writer = new BinaryWriter();
        writer.writeByte(this.mode);
        writer.writeInt32(this.methodId);

        if (this.mode === MODE_INVOKE || this.mode === MODE_RETURN) {
            writer.writeInt32(this.invokeId);
        }

        if (this.mode === MODE_RETURN) {
            writer.writeInt32(this.errId);
        }

        if (this.payload && this.payload.length > 0) {
            writer.buffers.push(this.payload);
        }

        const body = writer.getBytes();
        if (!includeFraming) return body;

        const header = Buffer.alloc(4);
        header.writeUInt32LE(body.length, 0);
        return Buffer.concat([header, body]);
    }

    static decode(buffer) {
        if (buffer.length < 5) throw new Error(`Packet too short: ${buffer.length}`);

        const mode = buffer.readUInt8(0);
        const methodId = buffer.readInt32LE(1);
        let offset = 5;
        let invokeId = 0;
        let errId = 0;

        if (mode === MODE_INVOKE || mode === MODE_RETURN) {
            invokeId = buffer.readInt32LE(offset);
            offset += 4;
        }

        if (mode === MODE_RETURN) {
            errId = buffer.readInt32LE(offset);
            offset += 4;
        }

        const payload = buffer.subarray(offset);
        return new RpcPacket(mode, methodId, invokeId, errId, payload);
    }
}

module.exports = {
    MODE_INVOKE,
    MODE_RETURN,
    MODE_NOTIFY,
    MARK_NULL,
    MARK_COMMON,
    BinaryWriter,
    BinaryReader,
    RpcPacket,
};
